### PR TITLE
Provide an option to drop vio state.

### PR DESCRIPTION
**Pull request by ppedrot**
This forbids turning such vio files into vo files, but at the same time it drastically reduces the amount of memory they need on disk.

This feature is accessible using `-quick-light` in `coqtop` and `coqc`, and there is a `quick-light` target in the `coq_makefile`. It has been asked repeatedly, amongst others by @charguer.

Currently, checking or compiling such a vio silently fails, but it is unrelated to this PR. It seems that any failure in vio compilation is caught at some point in the code, so this probably deserves a separate bug report.

### CODE REVIEW

File: tools/CoqMakefile.in [hunk 1]
```diff
@@ -380,9 +380,14 @@ optfiles: $(if $(DO_NATDYNLINK),$(CMXSFILES))
 .PHONY: optfiles
 
 # FIXME, see Ralph's bugreport
+quick: COQQUICK := -quick
 quick: $(VOFILES:.vo=.vio)
 .PHONY: quick
 
+quick-light: COQQUICK := -quick-light
+quick-light: $(VOFILES:.vo=.vio)
+.PHONY: quick
```
Comment (SkySkimmer) [inline tools/CoqMakefile.in]: Wrong .PHONY

Comment (charguer) [reply tools/CoqMakefile.in]: - I fear that it would be reduce the usability of the system to reuse the same extension .vio. Depending on whether -quick or -quick-light is used, the Makefile rules for building .vo files would be different, this is unsatisfying. Why not introduce a ".voi" extension, describing "compiled interface". Remark: in the future, the implementation of the code that produce ".voi" might be completely different from that that produces ".vio" files, and in fact the latter might get dropped at some point. 

- In passing, I would recommand changing the name from "-quick-light" (which, out of context, is really not intuitive/informative), to "-output-voi" (or possibly just "-voi"). If the review process had existed in the past, I would have suggested "-quick" to be named "-output-vio", much to better reflect the intention. For comparison, consider the example: "latexmk -pdf foo.tex", to indicate pdf output.)

- It seems that there is a missing option, for indicating Coq whether it should load .vo or .vio files of dependencies. This is very important. The current behavior (based on timestamp) is inadequate, and makes it very hard for "make" to work properly. I thus suggest the coqc option "-load-voi" to indicate that coqc should load .voi and not .vo files for dependency. It should fail if it does not find .voi. It might appear as a best effort to load the .vo if it exists, but if the .vo is not sufficiently up to date then this is doing more bad than good. Btw, using a different extension (voi vs vio) would also avoid changing the current behavior of .vio files, and avoid breaking the existing workflow.

- It is useful to make a distinction between "a .v file whose proofs have been checked against .vo dependencies" (producing a .vo file) and "a .v file whose proofs have been checked against .voi dependencies". The later should produce a ".vok" file, with empty contents, simply to indicate that the file has been successfully checked. The reason is that in the former case, universes have not been checked. It is useful (to avoid complete recompilation of an entire project) to be able to make coexist .vo and .vok files. It is sane to never produce a .vo file that would lack universe checks. In summary, the "-load-voi" option should produce a ".vok" file and not a ".vo" file as output.

- In the code I read "| BuildVio of bool"
=> I think that it would be more explicit to replace the boolean by named constructors, e.g. | VioWithProofs | VioWithoutProofs
=> Yet given what is written above, and the fact that vio might become deprecated one day (the day OCaml comes with proper multicore support), I would suggest to simply have "| BuildVio | BuildVoi" in the main datatype.

- Bonus to your patch: extend Coq_makefile generator to produce dependency rules of the form: "Foo.requires: Bar.voi", when Foo.v requires Bar.v.

Thanks

**Comment by ppedrot** [comment]
> I fear that it would be reduce the usability of the system to reuse the same extension .vio. Depending on whether -quick or -quick-light is used, the Makefile rules for building .vo files would be different, this is unsatisfying. Why not introduce a ".voi" extension, describing "compiled interface". Remark: in the future, the implementation of the code that produce ".voi" might be completely different from that that produces ".vio" files, and in fact the latter might get dropped at some point.

I actually mildly believe the exact opposite. I'd rather merge vo and vio than create a new kind of file. Having more potential dependencies looks like a combinatorial explosion in Makefile, and thus of the number of bugs. I personally have no idea how to fix the Makefile to be clever about which file to pick, and it looks to me like a really hard problem.

**Comment by charguer** [comment]
[my comment seems to be gone? reposting]

-  I fear that it would be reduce the usability of the system to reuse the same extension .vio. Depending on whether -quick or -quick-light is used, the Makefile rules for building .vo files would be different, this is unsatisfying. Why not introduce a ".voi" extension, describing "compiled interface". Remark: in the future, the implementation of the code that produce ".voi" might be completely different from that that produces ".vio" files, and in fact the latter might get dropped at some point. 

-  In passing, I would recommand changing the name from "-quick-light" (which, out of context, is really not intuitive/informative), to "-output-voi" (or possibly just "-voi"). If the review process had existed in the past, I would have suggested "-quick" to be named "-output-vio", much to better reflect the intention. For comparison, consider the example: "latexmk -pdf foo.tex", to indicate pdf output.

-  It seems that there is a missing option, for indicating Coq whether it should load .vo or .vio files of dependencies. This is very important. The current behavior (based on timestamp) is inadequate, and makes it very hard for "make" to work properly. I thus suggest the coqc option "-load-voi" to indicate that coqc should load .voi and not .vo files for dependency. It should fail if it does not find .voi. It might appear as a best effort to load the .vo if it exists, but if the .vo is not sufficiently up to date then this is doing more bad than good. Btw, using a different extension (voi vs vio) would also avoid changing the current behavior of .vio files, and avoid breaking the existing workflow.

-  It is useful to make a distinction between "a .v file whose proofs have been checked against .vo dependencies" (producing a .vo file) and "a .v file whose proofs have been checked against .voi dependencies". The later should produce a ".vok" file, with empty contents, simply to indicate that the file has been successfully checked. The reason is that in the former case, universes have not been checked. It is useful (to avoid complete recompilation of an entire project) to be able to make coexist .vo and .vok files. It is sane to never produce a .vo file that would lack universe checks. In summary, the "-load-voi" option should produce a ".vok" file and not a ".vo" file as output.

-  In the code I read "| BuildVio of bool"
=> I think that it would be more explicit to replace the boolean by named constructors, e.g. | VioWithProofs | VioWithoutProofs
=> Yet given what is written above, and the fact that vio might become deprecated one day (the day OCaml comes with proper multicore support), I would suggest to simply have "| BuildVio | BuildVoi" in the main datatype.

-  Bonus to your patch: extend Coq_makefile generator to produce dependency rules of the form: "Foo.requires: Bar.voi", when Foo.v requires Bar.v.

**Comment by SkySkimmer** [comment]
>[my comment seems to be gone? reposting]

It's in the "outdated comment" fold because it's attached to an outdated diff bit.

**Comment by ppedrot** [comment]
Relevant to the current discussion: #5223.

**Comment by ejgallego** [comment]
Also #4123. I support @ppedrot 's view of merging `vo` and `.vio` files, however that requires deep changes in the current implementation. IMHO, a `vo` should be perfectly fine with holes and these should be perfectly "fillable" by incremental proof checking.

At some point we would indeed like to be able to update only the parts of a `vo` file that have changed in the `.v` file [assuming we figure out how to do it wrt universes etc...]

I think that indeed another worthwhile goal to pursue could be to stop `Future.t` leaking down to the kernel. Why should the kernel be aware of that implementation detail for delayed proofs? The kernel should be happy to take a hole [and thus reflect it in print assumptions], in the same say it takes assumptions or section variable.

My point of view is that it is the high-level proof scheduler who should use whatever delay mechanism seems fit [either an `Async` task queue, futures, or whatever] and the update the document as completion events arrive.

With regards to the current proposal thou, it could make sense to have indeed a different file format, at least from the point of view of the user.

By the way, during the WG I mentioned the work on incremental checking, here is in case someone is interested: https://www.cs.utexas.edu/~celik/papers/CelikETAL17iCoq.pdf http://cozy.ece.utexas.edu/icoq/

Cheers!

**Comment by ppedrot** [comment]
> I think that indeed another worthwhile goal to pursue could be to stop Future.t leaking down to the kernel. Why should the kernel be aware of that implementation detail for delayed proofs? The kernel should be happy to take a hole [and thus reflect it in print assumptions], in the same say it takes assumptions or section variable.

+1, although I'm not sure it is doable in a satisfying way.

> At some point we would indeed like to be able to update only the parts of a vo file that have changed in the .v file [assuming we figure out how to do it wrt universes etc...]

The whole parallel fairy tale is a total lie anyways, there is currently not a single guarantee about the behaviour of non-sequential compilation of the code. In full polymorphic mode (i.e. with explicit constraints) we could give a meaning to it, but I'm pretty sure no sane user would like to write out the constraints generated by a proof term. In the mean time, we try to make our users happy with good-enough unsound approximations.

**Comment by ejgallego** [comment]
> +1, although I'm not sure it is doable in a satisfying way.

I'd be very interested if some day you can share more of your thoughts on the problems you see [apart from universes] In particular I am curious as to why replacing a hole is bound to be different from `lambda-lifting` + application.

The worst that can happen is that filling a hole indeed triggers a re-typechecking of all the proofs that follow/depend on the update. In fact, maybe only new universe constraints need to be checed. Indeed, this is to be handled with care, but I don't see how `Future.t` doesn't already have this problem as you point out below in your answer.

> The whole parallel fairy tale is a total lie anyways, there is currently not a single guarantee about the behaviour of non-sequential compilation of the code.

Well, the fact is that we have a mechanism doing that. I guess a first step would be to understand what we are doing exactly. IIUC, it is not possible to prove `False` [apart form the UI bugs] in a proof document that is `finish'ed`. [Well, I wouldn't bet a bitcoin on it thou]

**Comment by ppedrot** [comment]
> I'd be very interested if some day you can share more of your thoughts on the problems you see [apart from universes]

Obviously, if you remove the main source of trouble, things are *waaaaaay* easier.

> IIUC, it is not possible to prove False [apart form the UI bugs] in a proof document that is finish'ed. [Well, I wouldn't bet a bitcoin on it thou]

I wouldn't bet a bitcoin on the unprovability of False even without resorting to parallel compilation...

**Comment by ejgallego** [comment]
> Obviously, if you remove the main source of trouble, things are waaaaaay easier.

Well, I was indeed trying to figure out other showstoppers, at the next WG we'd like to propose an evolution on how we handle the parallel stuff.

So who knows / where it is described how the current code works?

It seems to me that the main entry point is `Safe_typing.join_safe_enviroment`. Do we know what concrete problems that function has? Should we open a tracking bug?

**Comment by ejgallego** [comment]
By the way I approve of this PR even if I'd like to have a closer look at some nits.

**Comment by charguer** [comment]
I think that you are missing a fundamental point of what we (end-users) are trying to achieve here. We want to be able to (1) update a coq file in the dependency chain and get back to a working state as fast as possible by only recompiling the interfaces of the files in the chain, and (2) allow to compile all interfaces of all files then compile all proofs, with parallelism introduced via "make -j". For the latter, it is essential to distinguish the interface files (.vio or .voi) from the compiled files (.vo). If you are trying to merge the two and progressively refine the .vo files, then you'll get data races in the files, which would get updated with proof contents while at the same time being read for the purpose of compiling files that depends on them. 

In summary: you do need distinct files to store compiled interface and to store full proof contents.

**Comment by ejgallego** [comment]
> In summary: you do need distinct files to store compiled interface and to store full proof contents.

Umm, I am not sure I do agree in that what you propose is a problem, but admittedly the next step on the architecture is not yet finish and of course I hope we can present it at the next WG so we can all discuss.

In the model we are trying to sketch, `.vo` files are nothing more than proof / elaboration caches and no data races exists as the top level Coq proof scheduler is the single process that has access to them. I don't see a way `make` can work for incremental compilation, and "icoq" seems to support that view.

I am not a fan of interface-only files as they do bring their own set of problems, and so far I have only seen one compelling use case that is to bootstrap mutually-dependent modules as is done in Haskell. But such use case doesn't seem to be planned to have in Coq yet, so we could as well do with a single file AFAICS, but of course it should have proper "phase separation" as sped things up. On the other hand, note that as of today, there is not a notion of interface-only file, due to several of the issues pointed out.

File: library/library.ml [hunk 2]
```diff
@@ -644,10 +644,9 @@ let load_library_todo f =
   let tasks, _, _ = System.marshal_in_segment f ch in
   let (s5 : seg_proofs), _, _ = System.marshal_in_segment f ch in
   close_in ch;
-  if tasks = None then user_err ~hdr:"restart" (str"not a .vio file");
-  if s2 = None then user_err ~hdr:"restart" (str"not a .vio file");
-  if s3 = None then user_err ~hdr:"restart" (str"not a .vio file");
-  if pi3 (Option.get s2) then user_err ~hdr:"restart" (str"not a .vio file");
+  let err = str "not a compilable file" in
```
Comment (gares) [inline library/library.ml]: I think this message should be more explicit, who knows what a compilable file is

**Review by gares — CHANGES_REQUESTED** [review]
I think the file should not be named `.vio` (I think .vi is more appropriate since they are really the closest thing we have today to an interface).

**Comment by gares** [comment]
Today `.vo` and `.vio` have the same file format, so I quite don't understand some of the comments up there.

I think the right format would be a (possibly zipped) directory, with subdirectories corresponding to today's segments:
```
foo.vo/
foo.vo/libstack
foo.vo/proof_terms/*
foo.vo/tasks/*
foo.vo/univs_from_tasks/*
```
Today tasks and proof_terms are array IIRC, but here are just directories.
A task generates a proof and some extra universe constraints.
A .vo has no tasks, all proof_Terms and eventually some univs_from_tasks (if generated via v->vio->vo).
A .vio lacks some proofs, has the corresponding tasks.

The .vi kind of file this PR is generating simply lacks some proofs_terms but has no corresponding tasks.

Note that such a file format is compatible with incremental vio->vo translation, even parallel I guess.
Today the `-check-proofs` thinghy is not able to generate .vo files simply because it is hard to build one (you can easily map, but then reduce is not obvious). With this forma, reduce is just putting a file in the right dir.

**Comment by gares** [comment]
> It seems that there is a missing option, for indicating Coq whether it should load .vo or .vio files of dependencies. This is very important. The current behavior (based on timestamp) is inadequate, and makes it very hard for "make" to work properly. I thus suggest the coqc option "-load-voi" to indicate that coqc should load .voi and not .vo files for dependency. It should fail if it does not find .voi. It might appear as a best effort to load the .vo if it exists, but if the .vo is not sufficiently up to date then this is doing more bad than good.

I may not remember correctly what the current code does, but IMO the case where you load an outdated .vo just because the .vio is not there should never happen. Make builds .vio files in topological order, the a more recent .vio should always be there. Then you can turn .vio into .vo in parallel, but then loading a finished .vo or an unfinished .vio should make no differece.

I don't see how the generate the scenario that needs the flag. Could you elaborate?

**Comment by maximedenes** [comment]
> I think .vi is more appropriate

This was discussed many times before, and we wanted to avoid that in case someday we introduce source interface files, for which it would be the natural naming.

**Comment by gares** [comment]
> This was discussed many times before, and we wanted to avoid that in case someday we introduce source interface files, for which it would be the natural naming.

Right, but I think in the meanwhile the one (hem,.. you) supporting this idea realized that the choice made by OCaml to have .ml and .mli is bad.... so my stubbornness came out ;-)

**Comment by ejgallego** [comment]
> Today .vo and .vio have the same file format, so I quite don't understand some of the comments up there.

At least in my view was more of merging them. I think the basic structure is about right, however we may want to drop the task completely and instead require `.v` files to fill `.vo` holes. In icoq it seems that a digest of the definition is made from the ast, that may be too weaker thou [and we may want a digest of the full path, I really need to see what they do in-depth].

I am curious about the universes from tasks and how they are handled, the ITP paper doesn't seem to provide any detail. Where could we look to understand the current method? I think you had some slides? A tech report maybe too?

**Comment by gares** [comment]
> I am curious about the universes from tasks

A proof has type, roughly, `(term * constraints) Future.t` and a task is morally a marshaling of that.
When you run it I split the pair, place the term in the opaque_terms array (that is not loaded when you load a .vo file) and the constraints in the univs_from_tasks section (that is loaded when you load the vo, together with the libstack).

If a .v file is compiled to .vo directly, all universe constraints are in the libstack.
If a .v file is compiled in two steps, then the the universe constraints of the statements/definitions are in the libstack, while the ones coming from proofs are in the univs_from_tasks segment.

**Comment by charguer** [comment]
@ejgallego  I agree that the day where you have a proper cross-file incremental compilation system, then using compiled interface files will become useless. Yet, it is going to take several years before it actually works well. In the meantime, I (and others) would benefit from compiled interfaces.

**Comment by ejgallego** [comment]
@charguer oh sure, I am not opposed to this PR at all, but just wanted to record some of the ideas that we've been floating around. It is indeed very likely that the approach changes as soon as we start implementing it.

**Comment by charguer** [comment]
Here is one proposal to give flexibility to the user, and cover (and explain) the current behavior of the system. Coqc would allow the user to specify:
- what is the main input file to process
- what is the extension to consider for the dependencies
- what is be the extension of the output file to produce
- what should be processed: interfaces, or proofs, or both.

Examples (assuming foo.v requires bar.v):
(some options could be activated by default)

Standard compilation (current behavior): 
  `coqc foo.v -do-compile -load vo -output vo`
 
Preparation of a vio file (current behavior):
  `coqc foo.v -do-suspend-proofs -load vo,vio -output vio`
  
Processing of the proofs in vio file (current behavior):  
  `coqc foo.vio -do-check-proofs -load vo,vio -output vo`
  
Compilation of an interface (new):
  `coqc foo.v -do-interface -load voi -output voi`
  
Verification of the proofs (new):  
  `coqc foo.v -do-compile -load voi -output vok`

**Comment by fpottier** [comment]
I support the idea that we need a clear, deterministic specification of which files are read/written by Coq. We do not want to Coq choose between two files based on their modification dates, for instance, as that would be very fragile / difficult to trace and debug. Something along the lines of Arthur's proposal sounds good.

Also, I would like to support the model where coq is invoked in batch mode to process *one* file, sequentially, and the problem of handling dependencies, parallel builds, etc. is dealt with by the build system (which could be "make" or some other system). So, please do not lock users into an approach where Coq itself must be the build system.

Thanks!

**Comment by ejgallego** [comment]
> So, please do not lock users into an approach where Coq itself must be the build system.

There won't be any "lock-in", but however, as things stand today, it is looking very likely that if you wanna enjoy from parallel incremental proof-checking in the future, you will have to delegate the task to Coq itself.

**Comment by ejgallego** [comment]
> to Coq itself.

to be more specific, you'll have to rely on a proof-checking scheduler written in OCaml. Whether you want to see that as part of Coq itself or not it is a more open ended question IMO.

**Comment by ejgallego** [comment]
Such scheduler will analyze a Coq project globally, determine with parts need re-checking/re-elaboration, create a task list, and proceed to update the on-disk caches.

**Comment by ejgallego** [comment]
Note that having such scheduler is typically what IDEs want. The workflow of an IDE is typically "load a big project, then the user makes small changes to it". So indeed Coq is complex enough that you may want to relief the IDE from having to keep track of the dependencies itself. Instead, the IDE just relays the changes the user has made to Coq, and Coq updates the TODO list accordingly.

**Comment by fpottier** [comment]
@ejgallego, that is precisely my point: I would like to be able to use `make` (or other tools) to program my own parallel builds. I am willing to work at the granularity level of one file. I just need `coqc` to offer sufficient flexibility to do this.

**Comment by ppedrot** [comment]
@fpottier I understand your demand, but I think a flag allowing to choose which object file to pick has not the right level of granularity. I'm slightly afraid that people will quickly start asking for the same feature at the file level.

Anyways, the parallel compilation feature is way above my head, we'd probably better off writing a CEP centralizing the exact demands from the user, otherwise we'll end up having a broken system not covering the user needs either...

**Comment by ejgallego** [comment]
@fpottier as far as I can tell, `coqc` will continue to work as it works today, maybe with a single flag to instruct it not to check the proofs. In this model, the way you build your dep tree is up to you.

**Comment by ejgallego** [comment]
In fact, I do not understand how could `coqc` not offer sufficiently flexibility to work on a file-by-file basis, unless we remove it, but nobody has ever proposed that AFAICT.

**Comment by fpottier** [comment]
A single flag instructing Coq not to check the proofs would be fine with me.

File: toplevel/usage.ml [hunk 3]
```diff
@@ -49,6 +49,7 @@ let print_usage_channel co command =
 \n  -compile-verbose f.v   verbosely compile Coq file f.v (implies -batch)\
 \n  -o f.vo                use f.vo as the output file name\
 \n  -quick                 quickly compile .v files to .vio files (skip proofs)\
+\n  -quick-light           same as -quick but does not store proof tasks\
```
Comment (gares) [inline toplevel/usage.ml]: I think this line should say
`quickly compile .v files to .voi files (drop proofs)`

File: library/library.ml (hunk 2 above)
Comment (ejgallego) [reply library/library.ml]: I suggest "the information to check proofs is not present in the file, you need to recompile the original file again."

**Review by ejgallego — APPROVED** [review]
It looks good to me modulo @gares requests. I am not sure about the terminology thou, `quick` `quick-light`, it would be good if we could find more descriptive terms.

**Comment by ppedrot** [comment]
I think I'll actually close this PR. Generating `voi` files that work in a satisfactory way would need a complete rework of the build system. Otherwise we will just make the situation worse, by augmenting  the combinatorial complexity of possible dependency chains. The current patch is way too naive to solve this kind of problem, and merging a half-baked PR is completely counter-productive IMHO.

I stand my ground, the most correct solution to the dependency hell is to merge `vo` and `vio` files. The analogy with OCaml is not meaningful, because `voi` do not really behave like interfaces, there is no such thing as a linking phase in Coq. Two kind of object files already need a form of dynamic computation of dependences. Adding a third kind of object file is a promise of unforeseen dire consequences.

Anyways, this patch is a hack. Let's calmly write a CEP instead.

**Comment by ejgallego** [comment]
I agree with your comment, however I do think this PR is useful, even if you strip the makefile support and leave users to handle their compilation dep chain.

**Comment by charguer** [comment]
I don't understand the exponential dependency problem. There is no need to arbitrarily mix voi and visiting files.

Either the user needs a traditional, full vo based compilation. Or the user needs a maximally independent compilation, building voi (interfaces), then playing or checking the desired proofs. It is fine imo to assume that voi (interface)  files are never involved in the creation of vo files (unlike what is the case currently with vio files, if I recall well).

If you drop the pull request,  then what I will have to do next week is to maintain a working copy of this patch in order to use it for my daily work ;)

**Comment by ejgallego** [comment]
@charguer even if @ppedrot is not interested anymore on this PR, I am, and I guess you are, so I suggest we re-submit a modified proposal that can be included in 8.8.

Even if the PR indeed has some room for improvement, I think the feedback we shall gather from advanced users it is totally worth it.

**Comment by charguer** [comment]
@ppedrot I understand that for a project 100% in Coq you might want a fancy incremental compilation system based on a unique representation of "partially evaluated files". However, I'd like to point out that for projects also involving tools other than Coq, it is completely essential to have distinct files for compiled interfaces and for compiled proofs. 

In my case, I need a Makefile that involves:
- compilation of an OCaml program
- this program parses OCaml code and generates Coq file
- these Coq files are compiled together with other libraries
- From other Coq files, OCaml code is extracted
- Various OCaml programs are then compiled
- These programs are executed against test data.

I'd like fine-grained control and good degree of parallelization throughout this entire process.
I do not see how an incremental compilation system for Coq could be smoothly integrated in my Makefile. It seems to me that the only reasonable option is to have a separate file type for compiled interfaces. Hence my motivation for this PR. I'm happy to discuss implementation details if you're not happy with the ones that you proposed initially.

**Comment by ejgallego** [comment]
I'd personally discourage using the terminology "interface" here, as maybe this is what is making @ppedrot wary. I don't think Coq will have "interfaces" in the short or medium term, but files whose proofs where not checked for speed.

**Comment by ejgallego** [comment]
@charguer and @silene propose that we instead provide an option to drop proofs from `.vo` files. I think this sounds good to me and should be doable. Make should work on the following account: `vio` files depend on other `vio` files, and `vo` files depend on other `vo` files, so you are either in one world or in the other.

WDYT @ppedrot ?

**Comment by silene** [comment]
> @charguer and @silene propose that we instead provide an option to drop proofs from `.vo` files.

Do I? If so, I hope I was talking about `.vio` files rather than `.vo` files.

**Comment by charguer** [comment]
If I recall well (sorry if I didn't) your suggestion was to reuse the format of vo files for storing files compiled without their proof; these files would have another extension (voi or vio), so as to preserve the existing meaning and contents of the vo extension.

**Comment by silene** [comment]
> your suggestion was to reuse the format of vo files for storing files compiled without their proof; these files would have another extension (voi or vio)

Oh right, I was confused about this thing about an option that would be doing things. I indeed said that the `.vo` format already supports the separation between informative content and proofs. So, while I understand the intent of the `.vio` format for serialization between instances of `coqtop`, the `.vo` format should be sufficient to store a summary of the content of a `.v` file.
